src/views/facts.ts

```diff
@@ -0,0 +1,108 @@
+import * as vscode from 'vscode';
+
+import { ConnectionHandler } from '../handler';
+import { RequestType } from 'vscode-languageclient';
+
+class PuppetFact extends vscode.TreeItem {
+  constructor(
+    public readonly label: string,
+    private value: string,
+    public readonly collapsibleState: vscode.TreeItemCollapsibleState,
+    public readonly children?: Array<[string, PuppetFact]>
+  ) {
+    super(label, collapsibleState);
+    this.iconPath = vscode.ThemeIcon.File;
+    // $facts['os']['family']
+    this.command = {
+      command: 'puppet.InsertFact',
+      title: 'Insert Fact into File',
+      arguments: [`$facts['${this.label}']`],
+    };
+  }
+
+  get tooltip(): string {
+    return `${this.label}-${this.value}`;
+  }
+
+  get description(): string {
+    return this.value;
+  }
+}
+
+interface PuppetFactResponse {
+  facts: string;
+  error: string;
+}
+
+export class PuppetFactsProvider implements vscode.TreeDataProvider<PuppetFact> {
+  private elements: Array<[string, PuppetFact]> = [];
+  private _onDidChangeTreeData: vscode.EventEmitter<PuppetFact | undefined> = new vscode.EventEmitter<
+    PuppetFact | undefined
+  >();
+  readonly onDidChangeTreeData: vscode.Event<PuppetFact | undefined> = this._onDidChangeTreeData.event;
+
+  constructor(protected handler: ConnectionHandler) {
+    vscode.commands.registerCommand('puppet.refreshFacts', () => this.refresh());
+    vscode.commands.registerCommand('puppet.InsertFact', (factString) => {
```

Comment:
TODO: The problem with this is that the command is triggered on left click, not right click. So if you are scrolling and click on an element by habit, it will insert the fact wherever the cursor is in an open manifest. Need to think on the UX here

Reply:
can you register an autocomplete for values in the `$facts` hash instead?

Reply:
We have a hover provider, where you get the fact resolved when you hover over `$facts['foo']`, but don't have autocomplete yet. But! We can! It's just hard! Because puppet facts! Are! Hard!

Sorry. Might have gotten carried away there with exclamations.

After thinking this through, I am going to create an issue for inserting facts from the view and another for autocomplete for the `$facts` variable, and constrain this PR to getting the view working with the language server.

Thanks for the input! Opps, there I go again.

Reply:
We already autocomplete on the 1st level facts (https://github.com/puppetlabs/puppet-editor-services/blob/master/lib/puppet-languageserver/manifest/completion_provider.rb#L40-L45)

So `$facts['os']`   We don't do second or deeper level (e.g. `$facts['os']['name']`) (Because ASTs are hard)